File: scripts/algorithms/P/Path Crossing/Path Crossing.py
# Runtime: 55 ms (Top 32.81%) | Memory: 14.1 MB (Top 28.49%)
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        c = set()
        x,y = 0,0
        c.add((x,y))
        for i in path:
            if i == 'N':
                y+=1
            elif i == 'E':
                x+=1
            elif i == 'W':
                x-=1
            else:
                y-=1
            if (x,y) in c:
                return True
            else:
                c.add((x,y))
        return False

File: scripts/algorithms/P/Path Crossing/Path Crossing.cpp
// Runtime: 3 ms (Top 74.64%) | Memory: 6.9 MB (Top 45.17%)
class Solution {
public:
    bool isPathCrossing(string path) {

        set<pair<int, int>>st;

        int x=0,y=0;

        st.insert({0, 0});

        for(int i=0;i<path.length();i++){

            if(path[i]=='N'){
                x++;
            }

            else if(path[i]=='S'){
                x--;
            }

            else if(path[i]=='E'){
                y++;
            }

            else{
                y--;
            }

            //if pair find at any point, return true
            if(st.find({x,y}) != st.end()){
                return 1;
            }
            //insert the pair
            st.insert({x, y});
        }
        return 0;
    }
};